Sources/Workspace/Workspace.swift

```diff
@@ -398,44 +419,77 @@ public class Workspace {
-                    branch: branch)
+                    branch: branch))
+                return
             }
             if let revision = revision, !repo.exists(revision: revision) {
-                throw WorkspaceDiagnostics.RevisionDoesNotExist(
+                diagnostics.emit(WorkspaceDiagnostics.RevisionDoesNotExist(
                     dependencyURL: dependency.repository.url,
-                    revision: revision.identifier)
+                    revision: revision.identifier))
+                return
             }
 
-            try handle.cloneCheckout(to: destination, editable: true)
-            let workingRepo = try repositoryManager.provider.openCheckout(at: destination)
-            try workingRepo.checkout(revision: revision ?? checkoutState.revision)
-            // Checkout to the new branch if provided.
-            if let branch = checkoutBranch {
-                try workingRepo.checkout(newBranch: branch)
+            guard diagnostics.wrap({
+                try handle.cloneCheckout(to: destination, editable: true)
+            }) else {
+                return
+            }
+            
+            guard let workingRepo = diagnostics.wrap({
+                try repositoryManager.provider.openCheckout(at: destination)
+            }) else {
+                return
+            }
+            
+            guard diagnostics.wrap({
+                try workingRepo.checkout(revision: revision ?? checkoutState.revision)
+                
+                // Checkout to the new branch if provided.
+                if let branch = checkoutBranch {
+                    try workingRepo.checkout(newBranch: branch)
+                }
+            }) else {
+                return
             }
         }
 
         // For unmanaged dependencies, create the symlink under editables dir.
         if let path = path {
-            try fileSystem.createDirectory(editablesPath)
-            // FIXME: We need this to work with InMem file system too.
-            try createSymlink(
-                    editablesPath.appending(component: packageName),
-                    pointingAt: path,
-                    relative: false)
+            guard diagnostics.wrap({
+                try fileSystem.createDirectory(editablesPath)
+                // FIXME: We need this to work with InMem file system too.
+                try createSymlink(
+                        editablesPath.appending(component: packageName),
+                        pointingAt: path,
+                        relative: false)
+            }) else {
+                return
+            }
         }
 
         // Save the new state.
         managedDependencies[dependency.repository] = dependency.editedDependency(
             subpath: RelativePath(packageName), unmanagedPath: path)
-        try managedDependencies.saveState()
+        
+        guard diagnostics.wrap({ try managedDependencies.saveState() }) else {
+            return
+        }
```

Comment:
I think we're better off creating two methods for edit, one throwing and one non throwing... that will remove the annoying guards everywhere